controllers/sandbox_controller.go

```diff
@@ -90,6 +90,13 @@ func (r *SandboxReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 		return ctrl.Result{}, err
 	}
 
+	// Check if already marked as expired to avoid repeated operations, including cleanups
+	if sandboxMarkedExpired(sandbox) {
```

Comment:
This check is currently placed before the `DeletionTimestamp` check (lines 106-109). If the Sandbox resource has any finalizers (either now or added in the future), this early return will prevent the finalizers from being processed and removed, causing the resource deletion to hang indefinitely. Please move this check *after* the deletion timestamp check.